mtcpclient/reporting.go

```diff
@@ -27,43 +34,34 @@ func ReportConnectionsStatus(gc GroupOfConnections, intervalBetweenUpdates int)
 // routines. It initializes and returns the channel that will be used for these communications
-func StartBackgroundReporting(numberConnections int, rinterval int) (chan tcpclient.Connection, GroupOfConnections) {
+func StartBackgroundReporting(numberConnections int, rinterval int) (chan tcpclient.Connection, *GroupOfConnections) {
 	// A connection may report up to 3 messages: Dialing -> Established -> Closed
 	const maxMessagesWeMayGetPerConnection = 3
 	connStatusCh := make(chan tcpclient.Connection, numberConnections*maxMessagesWeMayGetPerConnection)
-	connStatusTracker := make(GroupOfConnections, numberConnections)
+	connStatusTracker := GroupOfConnections{
+		connections: make([]tcpclient.Connection, numberConnections),
+		metrics: gcMetrics{
+			maxConcurrentEstablished: 0,
+		},
+	}
 
 	go ReportConnectionsStatus(connStatusTracker, rinterval)
-	go collectConnectionsStatus(connStatusTracker, connStatusCh)
+	go collectConnectionsStatus(&connStatusTracker, connStatusCh)
 
-	return connStatusCh, connStatusTracker
+	return connStatusCh, &connStatusTracker
 }
 
 func FinalMetricsReport(gc GroupOfConnections) (output string) {
-	// Report for Established connections
-	// TODO: remove duplication - this and the next block are almost equal. The MR structure could have a common
-	//  String() representation
-	if gc.AtLeastOneConnectionEstablished() {
-		mr := gc.calculateMetricsReport(tcpclient.ConnectionEstablished)
-		output = "Response time stats for " + strconv.Itoa(mr.numberOfConnections) +
-			" established connections min/avg/max/dev = " +
-			mr.min.Truncate(time.Microsecond).String() + "/" +
-			mr.avg.Truncate(time.Microsecond).String() + "/" +
-			mr.max.Truncate(time.Microsecond).String() + "/" +
-			mr.stdDev.Truncate(time.Microsecond).String() + "\n"
+	// Report Established Connections
+	output += "--- tcpgoon execution statistics --- \n" +
+		"Total established connections: " +
+		strconv.Itoa(len(gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionEstablished, tcpclient.ConnectionClosed}))) + "\n" +
+		"Max concurrent established connections: " +
+		strconv.Itoa(gc.metrics.maxConcurrentEstablished) + "\n" +
+		"Number of established connections on closure: " +
+		strconv.Itoa(len(gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionEstablished}))) + "\n"
 
-	}
-
-	// Report for Errored connections
-	if gc.AtLeastOneConnectionInError() {
-		mr := gc.calculateMetricsReport(tcpclient.ConnectionError)
-		output += "Time to error stats for " + strconv.Itoa(mr.numberOfConnections) +
-			" failed connections min/avg/max/dev = " +
-			mr.min.Truncate(time.Microsecond).String() + "/" +
-			mr.avg.Truncate(time.Microsecond).String() + "/" +
-			mr.max.Truncate(time.Microsecond).String() + "/" +
-			mr.stdDev.Truncate(time.Microsecond).String() + "\n"
-	}
+	output += gc.pingStyleReport()
```

Comment:
you did well encapsulating this as a different function, but I would not move the responsibility of generating two reports in case of failures/OKs to this report, and keep this "if" in this method